sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala

```diff
@@ -1354,7 +1354,8 @@ class HiveDDLSuite
     val indexName = tabName + "_index"
     withTable(tabName) {
       // Spark SQL does not support creating index. Thus, we have to use Hive client.
-      val client = spark.sharedState.externalCatalog.asInstanceOf[HiveExternalCatalog].client
+      val client =
+        spark.sharedState.externalCatalog.unwrapped.asInstanceOf[HiveExternalCatalog].client
```

Comment:
maybe we should define a util method to get hive client from `SparkSession`

Reply:
Why are we passing the client using the catalog in the first place? Is this convenience, or is there a reason why we can't pass both separately? I don't think we should be doing this without a reason. I don't like needless casts.

Reply:
@gatorsmile, the problem is that Hive classes access the Hive client from the HiveExternalCatalog in several places. Why is it getting the client this way? Shouldn't that client get passed separately?

Reply:
We want to get rid of Hive dependency completely in the near future. Currently, in the source code, only HiveExternalCatalog needs to use/access the `client`. 

I might not get your point. Could you explain how to pass the client when we keep the client in `HiveExternalCatalog`?

Reply:
There are several places that access the `client`, like `HiveSessionStateBuilder` and `SaveAsHiveFile`. My point is that assuming the catalog is a `HiveExternalCatalog`, casting, and accessing a field of the catalog to get the client isn't a good way to pass the client to where it is needed. Why doesn't this code pass both `client` and `catalog` to these classes to avoid this problem?

I don't think it matters that Hive may be removed in the future. Unless removing it is going to happen in a week or two, I'm skeptical that it is going to be soon enough that we should ignore problems with the implementation.

Reply:
In the 2.0 release, we did a lot of work to refactor the previous Hive integration and move the client to HiveExternalCatalog. This is still the ongoing work. If you have time, please help this effort too!

Reply:
@gatorsmile, what is the reason for passing the client as a field of HiveExternalCatalog and not on its own? This requires casting the catalog to access the client, and there doesn't appear to be an obvious reason not to pass the client separately.

Given that this is a problem -- the proposed interface was returning a wrapped catalog -- I'm trying to understand why it is this way and whether it should be changed.

Reply:
The client is used for interacting with Hive metastore. Conceptually, it should be part of `HiveExternalCatalog`. If we want to pass it as a field of HiveExternalCatalog , we need to create the client in SharedState, which is in the core module instead of the hive module. Since we are getting rid of Hive from the code base, we do not want a Hive-speicific SharedState. Any better idea?

Reply:
> Conceptually, it should be part of HiveExternalCatalog.

Why? If this is true, then why is it used outside of the catalog?

As far as a Hive-specific shared state, that sounds like the "right" way to do this. As long as this doesn't leak into the API, it isn't a big problem. But, practices like this that require breaking abstractions by casting to a known concrete class just to avoid passing multiple variables makes Spark more brittle. It is great to hear we want to eliminate Hive, but that doesn't mean the Hive code shouldn't be properly maintained while it is still supported.

Reply:
We plan to get rid of `HiveSessionStateBuilder` and treat `Hive` tables as the regular data source tables. `Hive` should be the same as the other external data sources. For the other data sources, we will not have a shared state and session state. Hive metastore is just used for global metastore. In the future, it is also pluggable. Thus, it does not make sense to re-introduce Hive SharedState in the current stage.

Reply:
Ideally Hive should not be a first-class citizen in Spark, it's just a data source and a catalog, and nothing more. We want to narrow down the scope of hive usage in Spark, and keeping it only in `HiveExternalCatalog` is a good choice.

This is still an on-going effort, as @rdblue pointed out, there are still 2 places Spark uses Hive directly:
1. `HiveSessionStateBuilder`. It's mostly for the ADD JAR, once we move the ADD JAR functionality to `ExternalCatalog`, we can fix it
2. `SaveAsHiveFile`. It's the data source part so it should be allowed to use Hive there. One thing we can improve is the hive client reuse. We cast `HiveExternalCatalog` to get the existing hive client, maybe there is a better way to do it without the ugly casting.

Reply:
The issue is out of scope this PR. We can continue addressing the related issues in the future PRs.

Reply:
@gatorsmile, I agree that we don't need to block this PR on the problem, but could you create or post the issues that track solving these problems?